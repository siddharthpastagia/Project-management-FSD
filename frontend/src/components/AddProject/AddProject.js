import React, { useEffect, useState } from "react";
import {
  Container,
  Row,
  Col,
  FormControl,
  InputGroup,
  Button,
  ListGroup,
  Alert,
  Form,
  FormLabel
} from "react-bootstrap";
import { useFormik } from "formik";
import * as Yup from "yup";
import "./AddProject.scss";
import * as _ from "lodash";

import SearchModal from "../common/SearchModal";

import {
  getAllProject,
  addNewProject,
  getProjectById,
  updateProjectById,
  deleteProjectById,
  getAllUsers
} from "../../api/Api";
import moment from "moment";

export const AddProject = () => {
  const initialStartDate = formatDate(new Date());
  const initialEndDate = formatDate(getAfterDate(1));
  let [projectList, setProjectList] = useState([]);
  let [projectUpdated, setProjectUpdated] = useState(false);
  let [projects, setProjects] = useState([]);
  let [sortMode, setSortMode] = useState(false);
  let [editMode, setEditMode] = useState(false);
  let [projectId, setProjectId] = useState("");
  let [projectName, setProjectName] = useState("");

  let [startDate, setStartDate] = useState(initialStartDate);
  let [endDate, setEndDate] = useState(initialEndDate);
  let [dateRequired, setDateRequired] = useState(false);

  let [priority, setPriority] = useState("");
  let [statusMessage, setStatusMessage] = useState({
    show: false,
    message: "",
    variant: ""
  });
  let [startDateSort, setStartDateSort] = useState(false);
  let [endDateSort, setEndDateSort] = useState(false);
  let [prioritySort, setPrioritySort] = useState(false);

  const fetchAllProjects = async () => {
    try {
      setProjects(await getAllProject());
      setProjectList(await getAllProject());
      setProjectUpdated(false);
    } catch (err) {
      setStatusMessage({
        ...statusMessage,
        show: true,
        message: err,
        variant: "danger"
      });
    }
  };

  const handleSort = field => {
    setSortMode(true);
    const sortByField = _.sortBy(projects, field);
    setProjects(sortByField);
  };
  const deleteProject = async project => {
    try {
      const resp = await deleteProjectById(project);
      setProjectUpdated(true);
      setStatusMessage({
        ...statusMessage,
        show: true,
        message: resp.message,
        variant: "success"
      });
    } catch (err) {
      setStatusMessage({
        ...statusMessage,
        show: true,
        message: err,
        variant: "danger"
      });
    }
    resetFormState();
    formik.resetForm();
  };

  const getNumOfCompletedTask = task => {
    const length = task.filter(item => item.status === "Completed").length;
    return length;
  };

  const editProject = async project => {
    try {
      const resp = await getProjectById(project._id);
      setEditMode(true);
      setProjectId(resp._id);
      setProjectName(resp.projectName);
      setDateRequired(resp.dateRequired);
      setStartDate(
        resp.startDate ? formatDate(resp.startDate) : initialStartDate
      );
      setEndDate(resp.endDate ? formatDate(resp.endDate) : initialEndDate);
      setPriority(resp.priority);
      setManager(resp.manager);
    } catch (err) {}
  };

  //Call this function after setStatusMessag to autoHide alert message

  const resetFormState = () => {
    setProjectName("");
    setStartDate(initialStartDate);
    setEndDate(initialEndDate);
    setPriority("");
    setManager("");
    setDateRequired(false);
    setEditMode(false);
  };

  function getAfterDate(num) {
    return moment().add(num, "day");
  }

  function formatDate(date) {
    return moment(date).format("YYYY-MM-DD");
  }

  let [showManagerModal, setShowManagerModal] = useState(false);
  let [managerList, setManagerList] = useState([]);
  let [manager, setManager] = useState("");

  const onCloseManagerModal = val => {
    setShowManagerModal(false);
  };
  const searchManager = async () => {
    setManagerList(await getAllUsers());
    setShowManagerModal(true);
  };
  const onSearchManager = (data, setFieldValue) => {
    setFieldValue("manager", data);
  };

  useEffect(() => {
    fetchAllProjects();
  }, [projectUpdated]);

  const handleChange = e => {
    let newList = [];
    if (e.target.value !== "") {
      newList = projectList.filter(item => {
        const pData = item.projectName.toLowerCase();
        const priorityData =
          item.priority !== null ? item.priority.toString().toLowerCase() : "";
        const startDateData = formatDate(item.startDate).toLowerCase();
        const endDateData = formatDate(item.endDate).toLowerCase();
        const filter = e.target.value.toLowerCase();
        return (
          pData.includes(filter) ||
          priorityData.includes(filter) ||
          startDateData.includes(filter) ||
          endDateData.includes(filter)
        );
      });
    } else {
      newList = projectList;
    }
    setProjects(newList);
  };
  const formik = useFormik({
    enableReinitialize: true,
    initialValues: {
      projectId: projectId,
      projectName: projectName,
      startDate: startDate,
      endDate: endDate,
      priority: priority,
      dateRequired: dateRequired,
      manager: manager
    },
    validationSchema: Yup.object({
      projectName: Yup.string().required("Please enter Project Name"),
      startDate: Yup.date().when("dateRequired", {
        is: true,
        then: Yup.date().required("Please enter start date"),
        otherwise: Yup.date()
      }),
      endDate: Yup.date().when("dateRequired", {
        is: true,
        then: Yup.date()
          .required("Please enter end date")
          .test(
            "compare",
            "End date should be greater than Start date",
            function(endDate) {
              return endDate > this.parent.startDate;
            }
          ),
        otherwise: Yup.date()
      }),
      priority: Yup.number().required("Please enter Priority"),
      dateRequired: Yup.boolean(),
      manager: Yup.object().required("Please select Manager from serach")
    }),
    onSubmit: async value => {
      if (!value.dateRequired) {
        delete value.startDate;
        delete value.endDate;
      }
      try {
        const resp = !editMode
          ? await addNewProject(value)
          : await updateProjectById(value);
        setProjectUpdated(true);
        setStatusMessage({
          ...statusMessage,
          show: true,
          message: resp.message,
          variant: "success"
        });
      } catch (err) {
        setStatusMessage({
          ...statusMessage,
          show: true,
          message: err,
          variant: "danger"
        });
      }
      resetFormState();
      formik.resetForm();
    }
  });

  return (
    <>
      <Container>
        <Row>
          <Col>
            <h3 className="mt-3">Project Management</h3>
          </Col>
        </Row>
        <Row>
          <Col className="mt-3">
            <Alert
              variant={statusMessage.variant}
              show={statusMessage.show}
              onClose={() => {
                setStatusMessage({ ...statusMessage, show: false });
              }}
              dismissible
            >
              {statusMessage.message}
            </Alert>
            <form onSubmit={formik.handleSubmit}>
              <Form.Group>
                <Form.Label>Project Name</Form.Label>

                <FormControl
                  required
                  placeholder="Project Name"
                  name="projectName"
                  errors={formik.errors.projectName}
                  className={
                    formik.touched.projectName
                      ? formik.errors.projectName
                        ? "is-invalid mb-2 mt-2"
                        : "is-valid mb-2 mt-2"
                      : "mb-2 mt-2"
                  }
                  {...formik.getFieldProps("projectName")}
                ></FormControl>
                <FormControl.Feedback type="invalid">
                  {formik.errors.projectName}
                </FormControl.Feedback>
              </Form.Group>

              <Form.Check
                type="checkbox"
                name="dateRequired"
                value={formik.values.dateRequired}
                checked={formik.values.dateRequired}
                label="Set Start and End date"
                {...formik.getFieldProps("dateRequired")}
              />
              <Form.Label className="mt-2">Start Date</Form.Label>
              <FormControl
                placeholder="Start Date"
                name="startDate"
                type="date"
                disabled={!formik.values.dateRequired}
                required={false}
                errors={formik.errors.startDate}
                className={
                  formik.touched.startDate
                    ? formik.errors.startDate
                      ? "is-invalid mb-2 mt-2"
                      : "is-valid mb-2 mt-2"
                    : "mb-2 mt-2"
                }
                {...formik.getFieldProps("startDate")}
              ></FormControl>
              <FormControl.Feedback type="invalid">
                {formik.errors.startDate}
              </FormControl.Feedback>
              <Form.Label>End Date</Form.Label>
              <FormControl
                placeholder="End Date"
                name="endDate"
                type="date"
                disabled={!formik.values.dateRequired}
                required={false}
                errors={formik.errors.endDate}
                className={
                  formik.touched.endDate
                    ? formik.errors.endDate
                      ? "is-invalid mb-2 mt-2"
                      : "is-valid mb-2 mt-2"
                    : "mb-2 mt-2"
                }
                {...formik.getFieldProps("endDate")}
              ></FormControl>
              <FormControl.Feedback type="invalid">
                {formik.errors.endDate}
              </FormControl.Feedback>
              <Form.Label>Priority</Form.Label>
              <FormControl
                placeholder="priority"
                name="priority"
                type="range"
                min="0"
                max="30"
                step="1"
                data-testid="priorityField"
                errors={formik.errors.priority}
                className={
                  formik.touched.priority
                    ? formik.errors.priority
                      ? "is-invalid mb-2 mt-2"
                      : "is-valid mb-2 mt-2"
                    : "mb-2 mt-2"
                }
                {...formik.getFieldProps("priority")}
              ></FormControl>

              <FormControl.Feedback type="invalid">
                {formik.errors.priority}
              </FormControl.Feedback>
              <Form.Label>Serach Manager</Form.Label>
              <InputGroup>
                {formik.values.manager.firstName ? (
                  <FormLabel className="modal-label">{`${formik.values.manager.firstName} ${formik.values.manager.lastName}`}</FormLabel>
                ) : (
                  <FormLabel className="modal-label"></FormLabel>
                )}
                <FormControl
                  plaintext
                  required
                  readOnly
                  hidden
                  name="manager"
                  errors={formik.errors.manager}
                  className={
                    formik.touched.manager
                      ? formik.errors.manager
                        ? "is-invalid mb-2 mt-2"
                        : "is-valid mb-2 mt-2"
                      : "mb-2 mt-2"
                  }
                  {...formik.getFieldProps("manager")}
                ></FormControl>
                <InputGroup.Append>
                  <Button onClick={searchManager}>
                    <i className="fa fa-search"></i>
                  </Button>
                </InputGroup.Append>
              </InputGroup>
              <FormControl.Feedback type="invalid">
                {formik.errors.manager}
              </FormControl.Feedback>

              <div className="text-center mt-4 mb-4">
                <Button
                  data-testid="updateProject"
                  variant="primary"
                  disabled={!formik.isValid || !formik.dirty}
                  type="submit"
                  className="ml-2 mr-2"
                >
                  {editMode ? "Update" : "Add"}
                </Button>
                <Button
                  variant="outline-secondary"
                  onClick={() => {
                    resetFormState();
                    formik.resetForm();
                  }}
                >
                  Reset
                </Button>
              </div>
              <SearchModal
                id="searchManager"
                heading="Serach Manager"
                showModal={showManagerModal}
                onCloseModal={onCloseManagerModal}
                data={managerList}
                columns={[
                  { dataField: "_id", hidden: true },
                  { dataField: "firstName", text: "User List" },
                  { dataField: "lastName", hidden: true }
                ]}
                onSearch={data => {
                  onSearchManager(data, formik.setFieldValue);
                }}
              />
            </form>
          </Col>
        </Row>
        <hr />
        <Row>
          <Col xs={12} sm={6}>
            <FormControl
              placeholder="Search"
              onChange={handleChange}
              className="mb-4"
              name="search"
            />
          </Col>
          <Col xs={12} sm={6}>
            Sort By:
            <Button
              variant="outline-primary"
              className={sortMode && startDateSort ? "active ml-2" : "ml-2"}
              data-testid="sortStartDate"
              onClick={() => {
                handleSort("startDate");
                setStartDateSort(true);
                setEndDateSort(false);
                setPrioritySort(false);
              }}
            >
              Start Date
            </Button>
            <Button
              variant="outline-primary"
              className={sortMode && endDateSort ? "active ml-2" : "ml-2"}
              data-testid="sortEndDate"
              onClick={() => {
                handleSort("endDate");
                setStartDateSort(false);
                setEndDateSort(true);
                setPrioritySort(false);
              }}
            >
              End Date
            </Button>
            <Button
              variant="outline-primary"
              className={sortMode && prioritySort ? "active ml-2" : "ml-2"}
              data-testid="sortPriority"
              onClick={() => {
                handleSort("priority");
                setStartDateSort(false);
                setEndDateSort(false);
                setPrioritySort(true);
              }}
            >
              Priority
            </Button>
          </Col>
        </Row>
        <Row>
          <Col>
            {projects.length === 0 && (
              <Alert variant="warning">
                Projects are not available in database!!
              </Alert>
            )}
            <ListGroup>
              {projects.map(project => {
                return (
                  <ListGroup.Item className="project-list" key={project._id}>
                    <div>
                      <p>Project : {project.projectName}</p>
                      <div>
                        <span className="mr-2">
                          Start Date :{" "}
                          {project.startDate && formatDate(project.startDate)}
                        </span>
                        <span className="ml-5">
                          End Date :{" "}
                          {project.endDate && formatDate(project.endDate)}
                        </span>
                      </div>
                      <div>
                        <span className="mr-2">
                          No of tasks: {project.task.length}
                        </span>
                      </div>
                      <div>
                        <span className="mr-2">
                          Completed: {getNumOfCompletedTask(project.task)}
                        </span>
                      </div>
                    </div>
                    <ListGroup>
                      <ListGroup.Item>
                        <p>Priority : </p> <p>{project.priority}</p>
                      </ListGroup.Item>
                    </ListGroup>
                    <div>
                      <Button
                        variant="outline-primary"
                        onClick={() => editProject(project)}
                      >
                        Update
                      </Button>
                      <br />
                      <Button
                        variant="outline-secondary"
                        className="mt-2"
                        onClick={() => {
                          deleteProject(project);
                        }}
                      >
                        Suspend
                      </Button>
                    </div>
                  </ListGroup.Item>
                );
              })}
            </ListGroup>
          </Col>
        </Row>
      </Container>
    </>
  );
};
