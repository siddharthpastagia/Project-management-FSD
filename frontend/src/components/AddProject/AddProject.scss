.project-list {
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.projectDetails {
  width: 30%;
}
