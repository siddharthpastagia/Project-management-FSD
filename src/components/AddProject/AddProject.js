import React, { useState } from "react";
import Modal from "react-bootstrap/Modal";
import { Button } from "react-bootstrap";
import Table from "../common/Table";

export const AddProject = () => {
  const [projectModal, setProjectModal] = useState(false);
  const columns = [{ dataField: "name", text: "Project Name" }];
  const projects = [
    { name: "Project A", id: "1" },
    { name: "Project B", id: "2" },
    { name: "Project C", id: "3" },
    { name: "Project D", id: "4" },
    { name: "Project E", id: "5" }
  ];
  return (
    <>
      <Modal
        show={projectModal}
        onHide={() => {
          setProjectModal(false);
        }}
      >
        <Modal.Header closeButton>
          <Modal.Title>Projects</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Table columns={columns} data={projects} keyField="id" />
        </Modal.Body>
      </Modal>
      This is Add project screen
      <Button variant="outline-info" onClick={() => setProjectModal(true)}>
        Open Modal
      </Button>
    </>
  );
};
