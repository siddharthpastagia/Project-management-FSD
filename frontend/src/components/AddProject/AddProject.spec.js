import React from "react";
import { MemoryRouter as Router } from "react-router-dom";
import { render, cleanup, wait, fireEvent } from "@testing-library/react";
import { toBeDisabled, toBeEmpty } from "@testing-library/jest-dom";
expect.extend({ toBeDisabled, toBeEmpty });
import { AddProject } from "./AddProject";

import axios from "axios";
jest.mock("axios");

describe("Add Project Component", () => {
  const onSubmit = jest.fn();
  const fakeProjectName = "ProjectA";
  const fakePriority = 11;
  const fakeManager = "";

  let projectList = {
    data: [
      {
        _id: "5df532c1f576f77b78e0e6dc",
        projectName: "ProjectA",
        priority: 24,
        dateRequired: true,
        manager: {
          _id: "5df53177f576f77b78e0e6d9",
          firstName: "User01",
          lastName: "user01lastName",
          empId: 1
        },
        endDate: "2019-12-21T00:00:00.000Z",
        startDate: "2019-12-14T00:00:00.000Z",
        task: [
          {
            _id: "5df5343af576f77b78e0e6e0",
            project: "5df532c1f576f77b78e0e6dc",
            taskName: "Parent Task 1 - p1u1(edited)",
            isParentTask: true,
            __v: 0,
            status: "Completed",
            id: "5df5343af576f77b78e0e6e0"
          }
        ],
        id: "5df532c1f576f77b78e0e6dc"
      }
    ]
  };

  const updateProjectResp = {
    data: {
      _id: "5df532c1f576f77b78e0e6dc",
      projectName: "ProjectA",
      priority: 24,
      dateRequired: true,
      manager: {
        _id: "5df53177f576f77b78e0e6d9",
        firstName: "User01",
        lastName: "user01lastName",
        empId: 1
      },
      endDate: "2019-12-21T00:00:00.000Z",
      startDate: "2019-12-14T00:00:00.000Z",
      task: [
        {
          _id: "5df5343af576f77b78e0e6e0",
          project: "5df532c1f576f77b78e0e6dc",
          taskName: "Parent Task 1 - p1u1(edited)",
          isParentTask: true,
          __v: 0,
          status: "Completed",
          id: "5df5343af576f77b78e0e6e0"
        }
      ],
      id: "5df532c1f576f77b78e0e6dc"
    }
  };

  beforeEach(() => {
    axios.get.mockResolvedValue(projectList);
  });

  afterEach(cleanup);

  it("Should render AddProject component", async () => {
    const { container } = render(
      <Router>
        <AddProject />
      </Router>
    );
    await wait(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it("Get project List ", async () => {
    axios.get.mockResolvedValue({ data: [{}] });
    const { getByText } = render(
      <Router>
        <AddProject />
      </Router>
    );
    expect(getByText("Projects are not available in database!!"));
  });

  it("Should able to Search Manager", async () => {
    const { container, getByText, getByTestId } = render(
      <Router>
        <AddProject />
      </Router>
    );
    const serachManagerBtn = getByTestId("searchManagerBtn");
    expect(serachManagerBtn).not.toBeDisabled();
    fireEvent.click(serachManagerBtn);
  });

  it("Should able to Suspend Project ", async () => {
    const { getByText } = render(
      <Router>
        <AddProject />
      </Router>
    );
    await wait(() => {
      const suspendBtn = getByText("Suspend");
      expect(suspendBtn).not.toBeDisabled();
      fireEvent.click(suspendBtn);
    });
  });

  it("Should able to Update Project ", async () => {
    const { getByPlaceholderText, getByText, getByTestId } = render(
      <Router>
        <AddProject />
      </Router>
    );
    let projectName, priority;
    await wait(() => {
      projectName = getByPlaceholderText("Project Name");
      priority = getByTestId("priorityField");
      const updateBtn = getByText("Update");
      axios.get.mockResolvedValue(updateProjectResp);
      fireEvent.click(updateBtn);
    });

    const updateProjectBtn = getByTestId("updateProject");
    fireEvent.change(projectName, {
      target: { name: "projectName", value: fakeProjectName }
    });
    fireEvent.change(priority, {
      target: { name: "priority", value: fakePriority }
    });
    fireEvent.click(updateProjectBtn);
  });

  it("Should able to Reset Project Form ", async () => {
    const { getByText, getByPlaceholderText } = render(
      <Router>
        <AddProject />
      </Router>
    );
    const projectName = getByPlaceholderText("Project Name");
    fireEvent.change(projectName, {
      target: { name: "projectName", value: fakeProjectName }
    });
    const resetBtn = getByText("Reset");
    expect(resetBtn).not.toBeDisabled();
    fireEvent.click(resetBtn);
    expect(projectName).toBeEmpty();
  });

  it("Should able to Serach the project from project List ", async () => {
    const { container, getByText, getByPlaceholderText } = render(
      <Router>
        <AddProject />
      </Router>
    );
    await wait(() => {
      const searchInp = getByPlaceholderText("Search");
      fireEvent.change(searchInp, {
        target: { name: "search", value: "ProjectA" }
      });
    });
    expect(getByText("Project : ProjectA"));
  });
  it("Should able to click on sorting options ", async () => {
    const { getByText, getByTestId } = render(
      <Router>
        <AddProject />
      </Router>
    );
    await wait(() => {
      const sortStartDateBtn = getByTestId("sortStartDate");
      const sortEndDateBtn = getByTestId("sortEndDate");
      const sortPriorityBtn = getByTestId("sortPriority");

      fireEvent.click(sortStartDateBtn);
      fireEvent.click(sortEndDateBtn);
      fireEvent.click(sortPriorityBtn);
      expect(sortStartDateBtn).not.toBeDisabled();
    });
    expect(getByText("Project : ProjectA"));
  });
});
