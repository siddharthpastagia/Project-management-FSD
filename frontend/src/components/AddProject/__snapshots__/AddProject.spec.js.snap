// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Project Component Should render AddProject component 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <h3
          class="mt-3"
        >
          Project Management
        </h3>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="mt-3 col"
      >
        <form>
          <div
            class="form-group"
          >
            <label
              class="form-label"
            >
              Project Name
            </label>
            <input
              class="mb-2 mt-2 form-control"
              name="projectName"
              placeholder="Project Name"
              required=""
              value=""
            />
            <div
              class="invalid-feedback"
            />
          </div>
          <div
            class="form-check"
          >
            <input
              class="form-check-input"
              name="dateRequired"
              type="checkbox"
              value="false"
            />
            <label
              class="form-check-label"
              title=""
            >
              Set Start and End date
            </label>
          </div>
          <label
            class="mt-2 form-label"
          >
            Start Date
          </label>
          <input
            class="mb-2 mt-2 form-control"
            disabled=""
            name="startDate"
            placeholder="Start Date"
            type="date"
            value="2019-12-15"
          />
          <div
            class="invalid-feedback"
          />
          <label
            class="form-label"
          >
            End Date
          </label>
          <input
            class="mb-2 mt-2 form-control"
            disabled=""
            name="endDate"
            placeholder="End Date"
            type="date"
            value="2019-12-16"
          />
          <div
            class="invalid-feedback"
          />
          <label
            class="form-label"
          >
            Priority
          </label>
          <input
            class="mb-2 mt-2 form-control"
            data-testid="priorityField"
            max="30"
            min="0"
            name="priority"
            placeholder="priority"
            step="1"
            type="range"
            value=""
          />
          <div
            class="invalid-feedback"
          />
          <label
            class="form-label"
          >
            Serach Manager
          </label>
          <div
            class="input-group"
          >
            <label
              class="modal-label form-label"
            />
            <input
              class="mb-2 mt-2 form-control-plaintext"
              hidden=""
              name="manager"
              readonly=""
              required=""
              value=""
            />
            <div
              class="input-group-append"
            >
              <button
                class="btn btn-primary"
                type="button"
              >
                <i
                  class="fa fa-search"
                />
              </button>
            </div>
          </div>
          <div
            class="invalid-feedback"
          />
          <div
            class="text-center mt-4 mb-4"
          >
            <button
              class="ml-2 mr-2 btn btn-primary"
              data-testid="updateProject"
              disabled=""
              type="submit"
            >
              Add
            </button>
            <button
              class="btn btn-outline-secondary"
              type="button"
            >
              Reset
            </button>
          </div>
        </form>
      </div>
    </div>
    <hr />
    <div
      class="row"
    >
      <div
        class="col-sm-6 col-12"
      >
        <input
          class="mb-4 form-control"
          name="search"
          placeholder="Search"
        />
      </div>
      <div
        class="col-sm-6 col-12"
      >
        Sort By:
        <button
          class="ml-2 btn btn-outline-primary"
          data-testid="sortStartDate"
          type="button"
        >
          Start Date
        </button>
        <button
          class="ml-2 btn btn-outline-primary"
          data-testid="sortEndDate"
          type="button"
        >
          End Date
        </button>
        <button
          class="ml-2 btn btn-outline-primary"
          data-testid="sortPriority"
          type="button"
        >
          Priority
        </button>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="list-group"
        >
          <div
            class="project-list list-group-item"
          >
            <div>
              <p>
                Project : 
                ProjectA
              </p>
              <div>
                <span
                  class="mr-2"
                >
                  Start Date :
                   
                  2019-12-13
                </span>
                <span
                  class="ml-5"
                >
                  End Date :
                   
                  2019-12-20
                </span>
              </div>
              <div>
                <span
                  class="mr-2"
                >
                  No of tasks: 
                  1
                </span>
              </div>
              <div>
                <span
                  class="mr-2"
                >
                  Completed: 
                  1
                </span>
              </div>
            </div>
            <div
              class="list-group"
            >
              <div
                class="list-group-item"
              >
                <p>
                  Priority : 
                </p>
                 
                <p>
                  24
                </p>
              </div>
            </div>
            <div>
              <button
                class="btn btn-outline-primary"
                type="button"
              >
                Update
              </button>
              <br />
              <button
                class="mt-2 btn btn-outline-secondary"
                type="button"
              >
                Suspend
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
